lib/AST/Type.cpp

```diff
@@ -1243,6 +1243,51 @@ CanType TypeBase::getCanonicalType(GenericSignature *sig) {
   return sig->getCanonicalTypeInContext(this);
 }
 
+bool TypeBase::isCanonicalizable() const {
+  switch (getKind()) {
+#define SUGARED_TYPE(id, parent)                                               \
+  case TypeKind::id:                                                           \
+    return cast<id##Type>(this)->isCanonicalizable();
+#define TYPE(id, parent)
+#include "swift/AST/TypeNodes.def"
+  case TypeKind::Enum:
+  case TypeKind::Struct:
+  case TypeKind::Class:
+  case TypeKind::Protocol: {
+    auto nominalTyParent = cast<NominalType>(this)->getParent();
+    return !nominalTyParent || nominalTyParent->isCanonicalizable();
+  }
+  case TypeKind::Tuple: {
+    auto tupleTy = cast<TupleType>(this);
+    for (const TupleTypeElt &field : tupleTy->getElements()) {
+      if (field.getType().isNull() || !field.getType()->isCanonicalizable()) {
+        return false;
+      }
+    }
+    return true;
+  }
+  case TypeKind::DependentMember: {
+    auto dependent = cast<DependentMemberType>(this);
+    return dependent->getBase()->isCanonicalizable();
+  }
+  default:
+    return true;
+  }
+}
+
+bool SugarType::isCanonicalizable() const {
+  switch (getKind()) {
+  case TypeKind::ArraySlice:
+  case TypeKind::Optional:
+  case TypeKind::Dictionary:
+    // Canonicalizing these types requires the standard library be loaded
+    return Bits.SugarType.HasCachedType ||
+           (Context && bool(Context->getStdlibModule()));
```

Comment:
Instead can we just fix the relevant tests to either define Array/Optional/Dictionary types, or just use the real stdlib instead of -parse-stdlib?

Reply:
They're validation crash tests, so no.

Reply:
I'm not sure what you mean. Why can't we change the test? `-parse-stdlib` is not really a supported configuration.

Another option is for the type checker to check for the presence of these decls early on, and emit a fatal error if they're not present. It just doesn't seem worth the added complexity to try to recover when they're missing.

Reply:
All the tests in question are SIL validation tests, so `-parse-stdlib` isn't used, they just aren't importing `Swift` explicitly.

I'm hesitant to just import the it, because while its hard to say for sure, it looks like these were originally crashers because diagnostics code was canonicalizing types when it shouldn't have been, just like in this case. (041-swift-typebase-getdesugaredtype.sil especially seems like it originally asserted during type canonicalization)

I'll look into diagnosing the absence of `Optional`/`ArraySlice`/`Dictionary` instead, but I do have a couple of concerns:
- That won't be enough to prevent a crash if an Tuple type with non-typechecked elements makes it into a diagnostic somehow
- It's potentially a bit fragile if more sugar types get added in the future

These might be perfectly acceptable tradeoffs, just thought it was worth considering! Sorry for all the back and forth on this, I've been learning a lot of this as I go along and underestimated the complexity.

Reply:
We don't guarantee that the compiler won't crash on invalid SIL source. It would take a lot of work to get to that point. I don't think the SIL crashes in validation-tests/SIL/crashers have much value except for the ones that trigger 'real' type checker bugs.

Not canonicalizing problematic types would be difficult to guarantee since we canonicalize types all over the place in the type checker, name lookup, etc.

Another thing you can try is canonicalizing sugar types to ErrorType when the relevant decls are missing.

Reply:
Thanks for clarifying, sounds like I was partially misunderstood the purpose of these tests. I ended up adopting your suggestion to canonicalize to ErrorType, it cuts out a lot of the code that got added to this PR and it looks like everything is still passing.